From f0e57691cd152a54e83ff9fc155d3d786e10c65c Mon Sep 17 00:00:00 2001
From: Tony Liu <Bo.Liu@windriver.com>
Date: Thu, 19 Aug 2010 23:05:54 -0700
Subject: [PATCH] Dynamic Ftrace: seperate code or data section for PowerPC module relocation

module_core is replaced by pax/grsec to module_core_rx and module_core_rw. So need to
distinguish them by section attribute.

Signed-off-by: Tonyliu <Bo.Liu@windriver.com>
---
 arch/powerpc/kernel/module_32.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/kernel/module_32.c b/arch/powerpc/kernel/module_32.c
index 0507238..96b8975 100644
--- a/arch/powerpc/kernel/module_32.c
+++ b/arch/powerpc/kernel/module_32.c
@@ -316,8 +316,14 @@ int apply_relocate_add(Elf32_Shdr *sechdrs,
 		}
 	}
 #ifdef CONFIG_DYNAMIC_FTRACE
+	if ((sechdrs->sh_flags & SHF_WRITE) || !(sechdrs->sh_flags & SHF_ALLOC))
 	module->arch.tramp =
-		do_plt_call(module->module_core,
+			do_plt_call(module->module_core_rw,
+				(unsigned long)ftrace_caller,
+				sechdrs, module);
+	else
+		module->arch.tramp =
+			do_plt_call(module->module_core_rx,
 			    (unsigned long)ftrace_caller,
 			    sechdrs, module);
 #endif
-- 
1.6.5.2

